### PR TITLE
tt cartridge

### LINKED ISSUE #34: tt cartridge
To work with the "cartridge" we want to reuse the https://github.com/tarantool/cartridge-cli module, except the following commands:
* start
* stop
* status
* connect

and probably:
* build
* pack
* create

The following implementation is suggested:
Add "cartridge-cli" to "tt" as a submodule with some wrapper that will mask excluded commands and pass others. Exceptions must also be taken into account for the "help" command.

Commands:
* start
* stop
* status
* connect
must be updated in this task to support "cartridge" functionality.

### PR DESCRIPTION
**Pull request by 0x501D**
This patchset adds cartridge-cli commands to tt:
*  admin
*  bench
*  build
*  create
*  failover
*  pack
*  repair
*  replicasets

Closes #34

### CODE REVIEW

File: cli/cartridge/extra/001_make_cmd_public.patch [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ 	// FLAGS
+diff --git a/cli/commands/connect.go b/cli/commands/connect.go
+index 7eebcd6..26c8c42 100644
+--- a/cli/commands/connect.go
++++ b/cli/commands/connect.go
+@@ -6,6 +6,11 @@ import (
+ 	"github.com/tarantool/cartridge-cli/cli/connect"
+ )
+ 
++var (
```
Comment (LeonidVas) [inline cli/cartridge/extra/001_make_cmd_public.patch:93]: AFAIU commands `Connect` and `Enter` we do not export, why do we create these variables?

File: cli/cartridge/extra/002_fix_admin_param.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/cli/commands/admin.go b/cli/commands/admin.go
+index 77fdb4e..9a9ed38 100644
+--- a/cli/commands/admin.go
++++ b/cli/commands/admin.go
+@@ -46,7 +46,7 @@ func addAdminFlags(flagSet *pflag.FlagSet) {
+ 	flagSet.BoolVarP(&ctx.Admin.Help, "help", "h", false, "Help for admin function")
+ 
+ 	flagSet.StringVar(&ctx.Admin.InstanceName, "instance", "", "Instance to connect to")
+-	flagSet.StringVarP(&ctx.Admin.ConnString, "conn", "c", "", "Address to connect to")
```
Comment (LeonidVas) [inline cli/cartridge/extra/002_fix_admin_param.patch:9]: Cobra does not allow to give the same name to global flags and command flags?

File: magefile.go [hunk 3]
```diff
@@ -1,3 +1,4 @@
+//go:build mage
```
Comment (LeonidVas) [inline magefile.go:1]: Please tell me WTF?

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patchset.
* AFAIU commit `build: add cartridge-cli submodule` does not add any commands to `tt` and not `Closes #34`, it just adds a submodule. This functionality is added by the `tt: integrate cartridge cli` commit.
* Since nothing works without the `build: bump go version` commit, this commit should be the first in the patch set.
* `cartridge: fix help for long commands.` It seems that the more correct submodule for the commit title is `help`, not `cartridge`.

File: cli/cartridge/extra/001_make_cmd_public.patch [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
++var CartridgeCliClean *cobra.Command
++
+ func init() {
+ 	var cleanCmd = &cobra.Command{
+ 		Use:   "clean [INSTANCE_NAME...]",
+@@ -23,6 +25,7 @@ func init() {
+ 		ValidArgsFunction: ShellCompRunningInstances,
+ 	}
+ 
++	CartridgeCliClean = cleanCmd
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/001_make_cmd_public.patch:81]: Moved to #79

File: cli/cartridge/extra/001_make_cmd_public.patch (hunk 1 above)
Comment (0x501D) [reply cli/cartridge/extra/001_make_cmd_public.patch:93]: Fixed

File: cli/cartridge/extra/002_fix_admin_param.patch (hunk 2 above)
Comment (0x501D) [reply cli/cartridge/extra/002_fix_admin_param.patch:9]: Yes. Cobra does not allow local flags to intersect with global ones.

File: go.mod [hunk 5]
```diff
@@ -2,6 +2,8 @@ module github.com/tarantool/tt
 
 go 1.17
 
+replace github.com/tarantool/cartridge-cli => ./cli/cartridge/third_party/cartridge-cli
```
[the remark this answers is not in the document]
Comment (0x501D) [reply go.mod:5]: fixed

File: magefile.go [hunk 6]
```diff
@@ -63,10 +66,73 @@ func init() {
...
+		panic(err)
+	}
+
+	err = os.Chdir(currDir)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Patch cartridge-cli.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply magefile.go:92]: Comments added.

File: magefile.go [hunk 7]
```diff
@@ -63,10 +66,73 @@ func init() {
 	os.Setenv("GO111MODULE", "on")
 }
 
+// Generate cartridge-cli Go code.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply magefile.go:69]: Comments added.

File: cli/cmd/help.go [hunk 8]
```diff
@@ -61,6 +61,11 @@ func getInternalHelpFunc(cmd *cobra.Command, help DefaultHelpFunc) modules.Inter
 		// 	is enough to check cmd.Name() == "tt".
 		// - `tt help -I`
 		// - `tt --help`, `tt -h` or `tt` (look code above).
+
+		// FIXME Some long commands from cartridge cli do not fall into modulesInfo.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/help.go:65]: Fixed

File: cli/cmd/help.go [hunk 9]
```diff
@@ -61,6 +61,11 @@ func getInternalHelpFunc(cmd *cobra.Command, help DefaultHelpFunc) modules.Inter
 		// 	is enough to check cmd.Name() == "tt".
 		// - `tt help -I`
 		// - `tt --help`, `tt -h` or `tt` (look code above).
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/help.go:63]: Fixed

**Pull request merged**